pocs/observatory.py

```diff
@@ -611,6 +618,156 @@ def close_dome(self):
             self.logger.info('Closed dome')
         return self.dome.close()
 
+    def take_flat_fields(self,
+                         which='evening',
+                         alt=None,
+                         az=None,
+                         min_counts=1000,
+                         max_counts=12000,
+                         target_adu_percentage=0.5,
+                         initial_exptime=3.,
+                         max_exptime=120.,
+                         camera_list=None,
+                         bias=2048,
+                         max_num_exposures=10,
+                         ):
+        """Take flat fields.
+        This method will slew the mount to the given AltAz coordinates(which
+        should be roughly opposite of the setting sun) and then begin the flat-field
+        procedure. The first image starts with a simple 1 second exposure and
+        after each image is taken the average counts are analyzed and the exposure
+        time is adjusted to try to keep the counts close to `target_adu_percentage`
+        of the `(max_counts + min_counts) - bias`.
+        The next exposure time is calculated as:
+            ```
+                exp_time = int(previous_exp_time * (target_adu / counts) *
+                           (2.0 ** (elapsed_time / 180.0)) + 0.5)
+            ```
+            Under - and over-exposed images are rejected. If image is saturated with
+            a short exposure the method will wait 60 seconds before beginning next
+            exposure.
+            Optionally, the method can also take dark exposures of equal exposure
+            time to each flat-field image.
+        Args:
+            which (str, optional): Specify either 'evening' or 'morning' to lookup coordinates
+                in config, default 'evening'.
+            alt (float, optional): Altitude for flats, default None.
+            az (float, optional): Azimuth for flats, default None.
+            min_counts (int, optional): Minimum ADU count.
+            max_counts (int, optional): Maximum ADU count.
+            target_adu_percentage (float, optional): Exposure time will be adjust so
+                that counts are close to: target * (`min_counts` + `max_counts`). Defaults
+                to 0.5.
+            initial_exptime (float, optional): Start the flat fields with this exposure
+                time, default 1 second.
+            max_exptime (float, optional): Maximum exposure time before stopping.
+            camera_list (list, optional): List of cameras to use for flat-fielding.
+            bias (int, optional): Default bias for the cameras.
+            max_num_exposures (int, optional): Maximum number of flats to take.
+        """
+        target_adu = target_adu_percentage * (min_counts + max_counts)
+        flat_obs = self._create_flat_field_observation(
+            which=which, alt=alt, az=az,
```

Comment:
@AnthonyHorton this might also be a question for you in regards to Huntsman. For PANOPTES we want to turn off the tracking during flat-fielding and it seems like we should do the same for Huntsman. Any particular reason to create the field and track it versus just going to initial position? I believe we were seeing stars in the Huntsman flats as well.

Reply:
It's possible to do flat fielding either way, with a fixed position or tracking at sidereal rate. In both cases you need to median combine a set of dithered images to remove all traces of stars from the processed flat field images. With the short exposure times and short focal length of PANOPTES it probably doesn't make much difference as the star trails will be short, but with longer focal lengths (e.g. Huntsman) it tends to work better with tracking. Sharp star images subtract more cleanly than trailed stars.

Reply:
Sounds like it would be easiest to just incorporate the dithering that we already have for Huntsman.

Supporting a non-tracking mode would probably be a nice addition as well however. I'll add it to the overview (#732) for now.